From bee30cf6aac599133ae1e89c09618a225b842714 Mon Sep 17 00:00:00 2001
From: Michael Allwright <allsey87@gmail.com>
Date: Tue, 18 Aug 2015 16:35:36 +0200
Subject: [PATCH] Update OV5640 and dependencies in DT. 1. Added regulators 2.
 Use GPIO to control clock 3. Added muxes, gpio expanders etc

---
 arch/arm/boot/dts/omap4-duovero-bebot.dts | 121 ++++++++++++++++++++++++------
 1 file changed, 96 insertions(+), 25 deletions(-)

diff --git a/arch/arm/boot/dts/omap4-duovero-bebot.dts b/arch/arm/boot/dts/omap4-duovero-bebot.dts
index f9fb013..26cb0d2 100644
--- a/arch/arm/boot/dts/omap4-duovero-bebot.dts
+++ b/arch/arm/boot/dts/omap4-duovero-bebot.dts
@@ -55,6 +55,34 @@
 		};
 	};
 
+	switch_ov5640_1_dvdd: switch_ov5640_1_dvdd {
+		compatible = "regulator-fixed";
+		regulator-name = "switch_ov5640_1_dvdd";
+		regulator-min-microvolt = <1800000>;
+		regulator-max-microvolt = <1800000>;
+		gpio = <&ov5640_1_gpio 0 GPIO_ACTIVE_HIGH>;
+		enable-active-high;
+	};
+
+	switch_ov5640_1_avdd: switch_ov5640_1_avdd {
+		compatible = "regulator-fixed";
+		regulator-name = "switch_ov5640_1_avdd";
+		regulator-min-microvolt = <2800000>;
+		regulator-max-microvolt = <2800000>;
+		gpio = <&ov5640_1_gpio 1 GPIO_ACTIVE_HIGH>;
+		enable-active-high;
+	};
+
+	switch_ov5640_1_vled: switch_ov5640_1_vled {
+		compatible = "regulator-fixed";
+		regulator-name = "switch_ov5640_1_vled";
+		regulator-min-microvolt = <4000000>;
+		regulator-max-microvolt = <4000000>;
+		regulator-always-on;
+		gpio = <&ov5640_1_gpio 2 GPIO_ACTIVE_HIGH>;
+		enable-active-high;
+	};
+
 	ocp {
 		iss: iss@52000000 {
 			compatible = "ti,omap4-iss";
@@ -93,7 +121,7 @@
 					channel = "0";
 					clock-lanes = <0>;
 					data-lanes = <1>;
-					remote-endpoint = <&ov5640_1>;
+					remote-endpoint = <&ov5640_1_ep>;
 				};
 			};
 		};
@@ -105,7 +133,6 @@
 			&led_pins
 			&button_pins
 			&smsc_pins
-			&camera_pins
 	>;
 
 	led_pins: pinmux_led_pins {
@@ -120,15 +147,6 @@
 		>;
 	};
 
-	camera_pins: pinmux_camera_pins {
-		pinctrl-single,pins = <
-			OMAP4_IOPAD(0x0be, PIN_OUTPUT | MUX_MODE3)		/* cam_strobe.gpio_82 */
-			OMAP4_IOPAD(0x0c0, PIN_OUTPUT | MUX_MODE3)		/* cam_globalreset.gpio_83 */
-			OMAP4_IOPAD(0x19a, PIN_OUTPUT | MUX_MODE0)		/* fref_clk1_out */
-		>;
-	};
-
-
 	i2c2_pins: pinmux_i2c2_pins {
 		pinctrl-single,pins = <
 			OMAP4_IOPAD(0x126, PIN_INPUT_PULLUP | MUX_MODE0)	/* i2c2_scl */
@@ -186,24 +204,77 @@
 
 	clock-frequency = <100000>;
 
-	ov5640: camera@3c {
-		compatible = "omnivision,ov5640";
-		status = "ok";
-		reg = <0x3c>;
+	i2c-mux@70 {
+		compatible = "nxp,pca9542";
+		#address-cells = <1>;
+		#size-cells = <0>;
+		reg = <0x70>;
+
+		ov5640_1: i2c@0 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0>;
+
+			ov5640_1_gpio: pca9554@38 {
+				compatible = "nxp,pca9554";
+				reg = <0x38>;
+				#gpio-cells = <2>;
+				gpio-controller;
+			};
 
-		pwdn-gpios = <&gpio3 18 GPIO_ACTIVE_HIGH>;	/* gpio_82 - power down */
-		reset-gpios = <&gpio3 19 GPIO_ACTIVE_LOW>;	/* gpio_83 - reset */
+			ov5640_1_camera: ov5640@3c {
+				compatible = "omnivision,ov5640";
+				status = "ok";
+				reg = <0x3c>;
 
-		clocks = <&auxclk1_ck>;
-		
-		port {
-			ov5640_1: endpoint {
-				clock-lanes = <0>;
-				data-lanes = <1>;
-				remote-endpoint = <&csi21>;
+				pwdn-gpios = <&ov5640_1_gpio 5 GPIO_ACTIVE_HIGH>;
+				reset-gpios = <&ov5640_1_gpio 6 GPIO_ACTIVE_LOW>;
+
+				clock-gpios = <&ov5640_1_gpio 4 GPIO_ACTIVE_HIGH>;
+
+				avdd-supply = <&switch_ov5640_1_avdd>;
+				dvdd-supply = <&switch_ov5640_1_dvdd>;
+
+				port {
+					ov5640_1_ep: endpoint {
+						clock-lanes = <0>;
+						data-lanes = <1>;
+						remote-endpoint = <&csi21>;
+					};
+				};
 			};
+
+			ov5640_1_flash: pca9632@60 {
+		                compatible = "nxp,pca9632";
+		                #address-cells = <1>;
+		                #size-cells = <0>;
+		                reg = <0x60>;
+		                tl@0 {
+		                        label = "ov5640_1_flash_tl";
+		                        reg = <0>;
+		                };
+		                tr@1 {
+		                        label = "ov5640_1_flash_tr";
+		                        reg = <1>;
+		                };
+		                bl@2 {
+		                        label = "ov5640_1_flash_bl";
+		                        reg = <2>;
+		                };
+		                br@3 {
+		                        label = "ov5640_1_flash_br";
+		                        reg = <3>;
+				};
+	                };
 		};
-	};
+
+		i2c@1 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <1>;
+		};
+	};	
+
 };
 
 &mmc3 {
-- 
1.9.1

